### PR TITLE
perf(geospatial): optimize grid.intersect() with KD-tree + ConvexHull

### PR DESCRIPTION
**Pull request by aacovski**
## Summary
Optimize `grid.intersect()` performance across all grid types using KD-tree spatial indexing with pre-computed ConvexHull equations and fully vectorized structured grid with `searchsorted`. This is all on the assumption that grids will never be non-convex and always layered.

## Performance Improvements
[geospatial_performance_results.csv](https://github.com/user-attachments/files/23699712/geospatial_performance_results.csv)
~50-300x speedup

## Key Optimizations
1. **KD-tree + vertices indexing**: O(log n) candidate cell search using centroids + vertices
2. **Pre-computed ConvexHull equations**: Vectorized point-in-polygon testing via hyperplane equations
3. **Bounding box rejection**: Fast elimination before expensive polygon tests
4. **Deepcopy fix**: Disabled `_copy_cache` during index build to avoid bottleneck
5. **Vectorized result processing**: Eliminated Python loops in grid.intersect() methods

## Changes
- Add `flopy/utils/geospatial_index.py` with GeospatialIndex class
- Optimize VertexGrid.intersect() - 2D kdtree, 2D convexhull and layer search
- Optimize UnstructuredGrid.intersect() `varies_by_layers=False` - 2D kdtree, 2D convexhull and layer search
- Optimize UnstructuredGrid.intersect() `varies_by_layers=True` - 3D kdtree, 3D convexhull
- Optimize StructuredGrid.intersect() with vectorized row/col finding - searchsorted
- Add edge case tests for thin sliver cells, boundaries, corners
- GeospatialIndex now raises ValueError if passed a StructuredGrid
- All methods now return np.nan for outside points (previously StructuredGrid returned (None, nan, nan) with z, and GeospatialIndex.query_point returned None)
- Array queries return int64 when all points inside, float64 when nan values present
- Matches behavior of native VertexGrid.intersect() and UnstructuredGrid.intersect() on develop
- 1:1 point-to-cell mapping
- All methods guarantee same number of outputs as inputs
- Outside points get nan instead of being dropped or raising errors (with forgive=True)

## Test coverage
- ~38 tests covering return type consistency, tie-breaking, 3D layered grids, and boundary handling

## Regressions
- grids with less than 10 cells or points perform worse due to overhead associated with new methods, but execute in less than 0.5 ms so it doesn't matter.

### CODE REVIEW

**Comment by aacovski** [comment]
**Boundary point dilemma:** DIS and DISV have different tie-breaking behavior for points on cell boundaries. The new unified method cannot match both simultaneously. We must either:
- Relax equivalence tests for boundary points (accept any valid adjacent cell)
- Choose one method to match exactly (fail tests for the other)

**Comment by aacovski** [comment]
@wpbonelli I’m looking for some guidance on how to handle these tests. The issue arises when a point lies exactly on the vertices of grid cells, making it ambiguous which of the 3-n cells it should belong to:

`FAILED test_grid.py::test_intersection - assert 268 == 267`

In these non-deterministic cases, different methods return different cell indices. I can force the kdtree-based methods to return the expected value, but that comes with a performance cost. It then causes another test failure (assert DIS == DISV), since those methods also behave differently. I could also update the structured grid path to enforce consistent behavior so all tests pass.

Before making those changes, I’d like your input on the preferred approach to resolve these cascading test failures.

**Comment by aacovski** [comment]
No performance hit with changing tiebreaking

**Comment by dbrakenhoff** [comment]
> @wpbonelli I’m looking for some guidance on how to handle these tests. The issue arises when a point lies exactly on the vertices of grid cells, making it ambiguous which of the 3-n cells it should belong to:

@aacovski in the gridintersect module, in case of ties, we decided to always return the lowest grid cell number. But I think that any approach as long as it is clearly documented (and maybe consistent across runs) should be fine?

**Comment by wpbonelli** [comment]
Agree with @dbrakenhoff I think lowest number is the existing convention so might as well stick to it if there is no performance cost?

**Comment by aacovski** [comment]
Great, that's what I did. I'll consider adding more tests for coverage, but otherwise I think the PR is ready. Thanks

**Comment by wpbonelli** [comment]
> across all grid types

> grids will never be non-convex and **always layered**

does the latter mean the new grid index and related performance boost don't support DISU?

**Comment by aacovski** [comment]
> does the latter mean the new grid index and related performance boost don't support DISU?

As I understand it, “fully 3D unstructured” grids are more characteristic of FEM models, and I’m not aware of MODFLOW supporting that kind of discretization. That said, my code does support DISU in the sense that `varies_by_layers` can be set to either `True` or `False`... I'm not aware of any FDM or FEM software supporting non-convex elemnts/grids. Please let me know if I'm missing anything.

**Comment by aacovski** [comment]
I am going to clean up docstrings later today as I noticed some old info. Sorry about that. I want to add some more tests too.

**Comment by wpbonelli** [comment]
@aacovski I guess I just didn't properly understand what you meant by "always layered"

File: autotest/test_geospatial_index.py [hunk 1]
```diff
@@ -0,0 +1,663 @@
...
+    nrow, ncol = 10, 10
+    delr = np.ones(ncol) * 10.0
+    delc = np.ones(nrow) * 10.0
+    top = np.ones((nrow, ncol)) * 10.0
+    botm = np.zeros((1, nrow, ncol))
+    return StructuredGrid(delr=delr, delc=delc, top=top, botm=botm)
+
+
+@pytest.fixture
+def layered_unstructured_grid():
```
Comment (wpbonelli) [inline autotest/test_geospatial_index.py]: If you think these simple/minimal grids might be generally useful for other tests, maybe consider moving them into `test_grid_cases.py` where we have collected a few already? Maybe in flopy 4.x we could consider providing example grids via some public API, similar to the way we [plan to provide example models](https://github.com/modflowpy/flopy/issues/1872) (prototype implementation [here](https://github.com/MODFLOW-ORG/modflow-devtools/issues/134))

File: autotest/test_geospatial_index.py [hunk 2]
```diff
@@ -0,0 +1,663 @@
...
+        ("minimal_vertex_grid", "vertex", 2, 5),
+        ("minimal_unstructured_grid", "unstructured", 2, 5),
+        ("simple_vertex_grid", "vertex", 100, 5),
+    ],
+)
+def test_index_build(
+    grid_fixture, grid_type, expected_cells, expected_points_per_cell, request
+):
+    """Test building index for different grid types."""
+    grid = request.getfixturevalue(grid_fixture)
```
Comment (wpbonelli) [inline autotest/test_geospatial_index.py]: nice. so you get named test cases with less repetition than parametrizing via the [object itself and `ids`](https://github.com/modflowpy/flopy/blob/e42f51117d1e0db8d9d7eb22eb8eae713a98262e/autotest/test_grid.py#L1332). TIL.

File: autotest/test_geospatial_index.py [hunk 3]
```diff
@@ -0,0 +1,663 @@
...
+    assert cellids[2] == 4
+    assert np.isnan(cellids[3])
+
+
+# ============================================================================
+# Test StructuredGrid Native Methods
+# ============================================================================
+
+
+def test_structured_boundary_tiebreaker(simple_structured_grid):
```
Comment (wpbonelli) [inline autotest/test_geospatial_index.py]: if this doesn't use `GeospatialIndex` under the hood, it probably belongs in `test_grid.py`?

File: autotest/test_geospatial_index.py [hunk 4]
```diff
@@ -0,0 +1,663 @@
...
+    result = index.query_points(x_mixed, y_mixed)
+    assert result.dtype == np.float64
+    assert not np.isnan(result[0])
+    assert np.isnan(result[1])
+
+    result = grid.intersect(x_mixed, y_mixed, forgive=True)
+    assert result.dtype == np.float64
+
+
+def test_return_type_structured(simple_structured_grid):
```
Comment (wpbonelli) [inline autotest/test_geospatial_index.py]: same as previous comment

File: autotest/test_geospatial_index.py [hunk 5]
```diff
@@ -0,0 +1,663 @@
...
+# ============================================================================
+# Test Edge Cases
+# ============================================================================
+
+
+def test_thin_sliver_cell():
+    """Test GeospatialIndex finds points in very thin "sliver" cells.
+
+    Tests the centroid+vertices KD-tree approach for cells where the
+    centroid might be far from the actual cell location.
```
Comment (wpbonelli) [inline autotest/test_geospatial_index.py]: what do you mean by "centroid might be far from the actual cell location"? the centroid of a triangular "sliver" cell will still be in the cell, right? wouldn't the cell have to be e.g. a crescent for the centroid to fall outside of it?

File: autotest/test_grid.py [hunk 6]
```diff
@@ -1655,20 +1655,3 @@ def test_unstructured_iverts_cleanup():
 
     if clean_ugrid.nvert != cleaned_vert_num:
         raise AssertionError("Improper number of vertices for cleaned 'shared' iverts")
-
-
-def test_structured_grid_intersect_edge_case(simple_structured_grid):
```
Comment (wpbonelli) [inline autotest/test_grid.py:1660]: as mentioned above I think the structured grid intersect tests should stay here since they don't use the geospatial index

**Review by wpbonelli — CHANGES_REQUESTED** [review]
it's maybe worth noting that the index assumes that x/y cell centers are in fact centroid coords? presumably the index requires this, rather than simple geometric mean, for polygons for which they aren't equivalent? maybe we should clarify in the docstrings on the grid classes as "cell centers" is ambiguous?

**Comment by aacovski** [comment]
> it's maybe worth noting that the index assumes that x/y cell centers are in fact centroid coords?

I hadn’t considered that the provided cell centers might not correspond to geometric centroids even for convex polygons. The index shouldn’t assume that, so I’ll update to clarify how cell centers are interpreted. Thanks for pointing it out!

File: autotest/test_geospatial_index.py (hunk 1 above)
Comment (aacovski) [reply autotest/test_geospatial_index.py]: I integrated the tests better into test_grid_cases.py and test_grid.py now.

**Comment by wpbonelli** [comment]
> The index shouldn’t assume that [the provided cell centers correspond to geometric centroids], so I’ll update to clarify how cell centers are interpreted.

Ah, ok- and the index doesn't even assume the center is inside the cell? It just performs better if so?

> I hadn’t considered that the provided cell centers might not correspond to geometric centroids even for convex polygons.

I think the intention was/is for it to be cell centroid, even though we don't check this when provided by the user or read from a binary grid file. Evidence:

- https://github.com/modflowpy/flopy/blob/e42f51117d1e0db8d9d7eb22eb8eae713a98262e/flopy/mf6/utils/binarygrid_util.py#L327 defines cell centers as centroids, maybe (but in any case suggestively) as a typo
  - https://github.com/modflowpy/flopy/blob/e42f51117d1e0db8d9d7eb22eb8eae713a98262e/flopy/mf6/utils/binarygrid_util.py#L329
- https://github.com/modflowpy/flopy/blob/e42f51117d1e0db8d9d7eb22eb8eae713a98262e/flopy/utils/cvfdutil.py#L17 is used when calculating x/y cell centers in other routines
  - https://github.com/modflowpy/flopy/blob/e42f51117d1e0db8d9d7eb22eb8eae713a98262e/flopy/utils/cvfdutil.py#L143
  - https://github.com/modflowpy/flopy/blob/e42f51117d1e0db8d9d7eb22eb8eae713a98262e/flopy/utils/cvfdutil.py#L306
  - https://github.com/modflowpy/flopy/blob/e42f51117d1e0db8d9d7eb22eb8eae713a98262e/flopy/utils/cvfdutil.py#L369
- https://github.com/modflowpy/flopy/blob/e42f51117d1e0db8d9d7eb22eb8eae713a98262e/flopy/utils/geometry.py#L739 is used from
  - https://github.com/modflowpy/flopy/blob/e42f51117d1e0db8d9d7eb22eb8eae713a98262e/flopy/discretization/unstructuredgrid.py#L1062

Maybe something to clarify in the MF6IO guide. For DIS it doesn't matter but it becomes relevant for DISV/U.

**Comment by wpbonelli** [comment]
@aacovski it wasn't really clear from previous comment but I think it's best to revert the docstrings to say centroid rather than leaving it ambiguous. my first comment was really to point out the (unwanted?) ambiguity in the existing grid classes, not to say your new ones should change

**Comment by wpbonelli** [comment]
Another thought: why not move the structured grid searchsorted query implementation to `GeospatialIndex`? That way all the `Grid.intersect` methods could use the index. Since the index receives the grid on init, it knows what kind it is. This seems a bit neater, and also opens the way for cases where a structured grid query might still find the KD tree useful, e.g. ball queries.

Also just want to say, thanks for working on this. This seems like it will be a great asset both in the near term and longer term- in 4.x it could sit behind xarray-based user APIs, whether builtin `.sel` syntax (point queries) or, for the fancy stuff, maybe an accessor.

**Comment by aacovski** [comment]
I'll work on implementing that: moving intersect to base class and remove the subclass intersect methods. Regarding the docstrings: since there is a difference between centroid and centers, based on everything I've read, it makes more sense to use the more general case (centers) rather than centroids which could be inaccurate.

**Comment by wpbonelli** [comment]
I'm pretty sure "cell centers" is meant as centroids, that's why I'm suggesting to change the doctrings up in the grids themselves. But I could be wrong. Hopefully someone with more tenure can chime in.

**Comment by christianlangevin** [comment]
> I'm pretty sure "cell centers" is meant as centroids, that's why I'm suggesting to change the doctrings up in the grids themselves. But I could be wrong. Hopefully someone with more tenure can chime in.

For DISV grids the user provides the x and y node coordinates for each cell.  In general, this coordinate is the centroid of the cell, however, there is at least one situation where this is not the case.  Along the edge of the model domain, there may be half cells (common with Voronoi) where it may be beneficial to assign the x,y cell center coordinate to be on the model domain boundary (and the edge of the cell).  The location of a constant-head boundary (or constant concentration), for example, would effectively be assigned to the node coordinate, which a user may want to put on the model edge.  This x, y node coordinate is used by DISV (and DISU in some cases) to calculate geometric properties for connections between adjacent cells.

**Comment by wpbonelli** [comment]
thanks @christianlangevin. so @aacovski the more general "cell centers" is better as you suggest. we may update the description upstream in mf6io but no need to change the grid docstrings here
